esp-hal-common/src/dma/mod.rs

```diff
@@ -802,3 +854,313 @@ pub trait DmaTransferRxTx<BR, BT, T>: Drop {
     /// Wait for the transfer to finish.
     fn wait(self) -> (BR, BT, T);
 }
+
+#[cfg(feature = "async")]
+pub(crate) mod asynch {
+    use core::task::Poll;
+
+    use super::*;
+    use crate::macros::interrupt;
+
+    pub struct DmaTxFuture<'a, TX> {
+        pub(crate) tx: &'a mut TX,
+        _a: (),
+    }
+
+    impl<'a, TX> DmaTxFuture<'a, TX>
+    where
+        TX: Tx,
+    {
+        pub fn new(tx: &'a mut TX) -> Self {
+            tx.listen_eof();
+            Self { tx, _a: () }
+        }
+    }
+
+    impl<'a, TX> core::future::Future for DmaTxFuture<'a, TX>
+    where
+        TX: Tx,
+    {
+        type Output = (); // TODO handle DMA errors
+
+        fn poll(
+            self: core::pin::Pin<&mut Self>,
+            cx: &mut core::task::Context<'_>,
+        ) -> Poll<Self::Output> {
+            TX::waker().register(cx.waker());
```

Comment:
@MabezDev Shouldn't the waker only be registered when the poll returns Pending? Since the waker overwrites the current one, this can register a waker that will maybe never wake. I wonder if this might be the cause of a heisenbug I experience where one of my tasks just stops working.

I know this is a common pattern in embassy but I wonder if it's correct.

Reply:
Perhaps? Did you try moving the waker registration? I can't really see it causing an issue, but if it fixes your issue its something to look into.

Reply:
No I'm working on another issue. I don't expect this to cause problems, just worse than ideal perf, so nothing terribly important.